test/gpu/native/studies/sort/radixSortGpu.chpl

```diff
@@ -0,0 +1,279 @@
+      //     if i >= 1 << d { // 2^d
+      //       prefixSum[i] = prefixSum[i-(1 << (d-1))] + prefixSum[i];
+      //     }
+      //   }
+      // }
+      hostPrefixSum=prefixSum;
+    }
+  }
+
+  proc blellochScan(ref hostprefixSum: [] uint, hostbuckets: int){
+      // Blelloch Scan
+      // This is the same as the count array, but each element is the sum of all previous elements
+      // Uses a more efficient algorithm that does O(n) work
+      // Blelloch (1990)
+      // Up-sweep
+      on here.gpus[1]{
+        var buckets = hostbuckets;
+        var prefixSum = hostprefixSum;
+        var offset = 1;
+        while offset < buckets {
+          var prefixSumBuffer = prefixSum;
+          foreach idx in 0..<buckets/(2*offset) {
+            const i = idx*2*offset;
+            prefixSum[i+2*offset-1] = prefixSumBuffer[i+offset-1] + prefixSumBuffer[i+2*offset-1];
+          }
+          offset = offset << 1;
+        }
+
+        // Down-sweep
+        prefixSum[buckets-1] = 0;
+        offset = buckets >> 1;
+        while offset > 0 {
+          var prefixSumBuffer = prefixSum;
+          foreach idx in 0..<buckets/(2*offset) {
+            const i = idx*2*offset;
+            const t = prefixSumBuffer[i+offset-1];
+            prefixSum[i+offset-1] = prefixSumBuffer[i+2*offset-1];
+            prefixSum[i+2*offset-1] = prefixSum[i+2*offset-1] + t;
+          }
+          offset = offset >> 1;
+        }
+        hostprefixSum = prefixSum;
+      }
+  }
+
+  proc arrScan(ref count: [?D] uint, ref inputArr: [] uint, exp: int, bitMask: int, buckets: int){
+    // Count the number of occurrences of each bitsAtATime-bit chunk in the array
+    // at the current exp position
+    on here.gpus[1]{
+      var gpuCount : [D] uint; // Initialized all 0s
+      var gpuInputArr = inputArr;
+      foreach i in gpuInputArr {
+        assertOnGpu();
+        gpuAtomicAdd(gpuCount[((i>>exp) & bitMask):int], 1);
+        // count[((i>>exp) & bitMask):int] += 1;
+      }
+      count = gpuCount;
+    }
+    if scanTest then
+      writeln(" Count: ", count);
+
+
+    if scanType == "hillisSteele" then
+      hillisSteeleScan(count, buckets);
+    else if scanType == "blelloch" then
+      blellochScan(count, buckets);
+    else if scanType == "serial" then
+      serialScan(count, buckets);
+
+    if scanTest then
+      writeln("Pf sum: ",count);
+  }
+
+  proc radixSort(inputArr: [] uint){
+    if (noisy){
+        writeln("Bits at a time: ", bitsAtATime);
+        writeln("Buckets: ", buckets);
+        writeln("Bit mask: ", bitMask);
+        writeln("Input: ", inputArr);
+    }
+
+    // Create the output array and copy input
+    var outputArr: [0..<inputArr.size] uint;
+
+    // Create the count and prefix sum arrays once
+    // This was we can reuse it for each iteration of radix Sort
+    var prefixSum: [0..<buckets] uint;
+
+    // Ceiling on number of iterations based on max element in array
+    // Can reduce do this faster?
```

Comment:
I certainly hope so, but reductions are a bit problematic with the GPU locale model. See https://github.com/chapel-lang/chapel/issues/22736. It maybe OK to leave it as is, but please link to this issue for the reason you are not doing that.

Reply:
I'll make that note and leave it as is for now